src/controller/CHIPDeviceController.cpp

```diff
@@ -1175,6 +1184,49 @@ void DeviceCommissioner::OnFailedToExtendedArmFailSafeDeviceAttestation(void * c
     commissioner->CommissioningStageComplete(CHIP_ERROR_INTERNAL, report);
 }
 
+void DeviceCommissioner::OnICDSymmetricKeyGenerationCompleted(void * context, CHIP_ERROR error, NodeId checkInNodeId,
+                                                              uint64_t subjectId, ICDRegistrationDelegate::ICDKey key)
+{
+    DeviceCommissioner * commissioner = static_cast<DeviceCommissioner *>(context);
+
+    if (error != CHIP_NO_ERROR)
+    {
+        ChipLogError(Controller, "Failed to generate symmetric key for ICD registration. Error %s", ErrorStr(error));
+        commissioner->CommissioningStageComplete(error);
+        return;
+    }
+    if (commissioner->mDeviceBeingCommissioned == nullptr)
+    {
+        ChipLogError(Controller, "No device is being commissioned at this time.");
+        commissioner->CommissioningStageComplete(CHIP_ERROR_INTERNAL);
+        return;
+    }
+    CommissioningDelegate::CommissioningReport report;
+    ICDRegistrationInfo info = ICDRegistrationInfo{
+        .checkInNodeId = checkInNodeId,
+        .subjectId     = subjectId,
+        .key           = key,
+    };
+    report.Set<ICDRegistrationInfo>(info);
+    commissioner->CommissioningStageComplete(CHIP_NO_ERROR, report);
+}
+
+void DeviceCommissioner::OnICDManagementRegisterClientResponse(
+    void * context, const app::Clusters::IcdManagement::Commands::RegisterClientResponse::DecodableType & data)
+{
+    DeviceCommissioner * commissioner = static_cast<DeviceCommissioner *>(context);
+
+    CommissioningDelegate::CommissioningReport report;
+    ICDRegistrationResponseInfo info;
+    info.icdCounter = data.ICDCounter;
+    report.Set<ICDRegistrationResponseInfo>(info);
+    commissioner->CommissioningStageComplete(CHIP_NO_ERROR, report);
```

Comment:
Shouldn't CommissioningStageComplete be after we notify `commissioner->mICDRegistrationDelegate`?